### PR TITLE
Add dashed-ident-no-missing-reference-function

### LINKED ISSUE #5000: Add dashed-ident-no-missing-reference-function
> What is the problem you're trying to solve?

Developers sometimes accidentally forget to include the `var()` function call when trying to dereference custom properties. For example, they might write:

```css
a {
  --my-color: --other-color;
  color: --my-color;
}
```

when they meant

```css
a {
  --my-color: var(--other-color);
  color: var(--my-color);
}
```

> What solution would you like to see?

A new rule that flags identifiers beginning with `--` that appear outside of the first argument of a `var()` function.

### PR DESCRIPTION
**Pull request by jeddy3**
> Which issue, if any, is this issue related to?

Closes #5000

> Is there anything in the PR that needs further explanation?

I wanted to put together something to help feed into the discussion in #5000.

I've dropped `declaration-value` from the name and went with `dashed-ident-no-missing-reference-function` as I now feel `declaration-value` is redundant because [_dashed-idents_](https://drafts.csswg.org/css-values-4/#dashed-idents) outside of declarations values don't use reference functions.

In the future, we can add `ignoreDashedIdents` and `ignoresFunction` options:

```json
"rules": {
  "dashed-ident-no-missing-reference-function": [true, {
    "ignoreDashedIdents": ["--my-js-ident"],
    "ignoreFunctions": ["--custom"]
  }]
}
```

Which would allow:

```css
a {
  prop: --my-js-ident;
  prop: --custom(--foo);
}
```

Like the [`font-family-no-missing-generic-family-keyword` rule](https://stylelint.io/user-guide/rules/font-family-no-missing-generic-family-keyword), this rule sits in the [possible errors category](https://stylelint.io/user-guide/rules/list#possible-errors).

### CODE REVIEW

File: lib/rules/dashed-ident-no-missing-reference-function/index.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+const validateOptions = require('../../utils/validateOptions');
+
+const ruleName = 'dashed-ident-no-missing-reference-function';
+
+const messages = ruleMessages(ruleName, {
+	rejected: (dashedIdent) =>
+		`Unexpected missing dashed-ident reference function for "${dashedIdent}"`,
+});
+
+const REFERENCE_FUNCTIONS = ['color', 'env', 'var'];
```
Comment (malsf21) [inline lib/rules/dashed-ident-no-missing-reference-function/index.js:19]: Curious - are there many other uses cases for this set of reference functions? Can't think of any in my normal use-cases, but thinking about if we should pull this out to the constants files.

**Review by malsf21 — APPROVED** [review]
Looking through the linked discussion, test cases, and implementation, this seems good to me! 

P.S. `a { color: calc(var(--foo) + --bar)); }` is a great test case for this to work on!

File: lib/rules/dashed-ident-no-missing-reference-function/index.js (hunk 1 above)
Comment (jeddy3) [reply lib/rules/dashed-ident-no-missing-reference-function/index.js]: I think we can pull them out into a reference file if they end up being shared by another rule.

**Comment by jeddy3** [comment]
Added a small commit to use `Set` to align with the changes in https://github.com/stylelint/stylelint/pull/5048.

File: lib/rules/dashed-ident-no-missing-reference-function/README.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# dash-ident-no-missing-reference-function
```
Comment (ybiquitous) [inline lib/rules/dashed-ident-no-missing-reference-function/README.md:1]: Typo 😅 

```suggestion
# dashed-ident-no-missing-reference-function
```

**Review by ybiquitous — APPROVED** [review]
I found one typo, but this change looks good to me! 👍

**Comment by alexander-akait** [comment]
Found this name is very misleading, in css we have `env` https://developer.mozilla.org/en-US/docs/Web/CSS/env() and in future them will be more

**Comment by jeddy3** [comment]
> Found this name is very misleading, in css we have env https://developer.mozilla.org/en-US/docs/Web/CSS/env() and in future them will be more

Yes, I had a similar thought. 

Opened https://github.com/stylelint/stylelint/pull/5317.

**Pull request closed without merging**